sphinx/builders/__init__.py

```diff
@@ -101,16 +106,45 @@ def __init__(self, app):
         self.parallel_ok = False
         self.finish_tasks = None  # type: Any
 
-        # load default translator class
-        self.translator_class = app._translators.get(self.name)
-
     def set_environment(self, env):
         # type: (BuildEnvironment) -> None
         """Store BuildEnvironment object."""
         self.env = env
         self.env.set_versioning_method(self.versioning_method,
                                        self.versioning_compare)
 
+    def get_translator_class(self, *args):
+        # type: (Any) -> nodes.NodeVisitor
+        """Return a class of translator."""
+        return self.app.factory.get_translator_class(self)
```

Comment:
In this case,  it seems that `app.factory`works as a registry for internal mechanism.
So every internal mechanism should access to `app.registry` instead of `app.factory` if my understanding is correct.
It means that app.factory has two responsibility; factory and registry.

BTW (not important) I feel that this reference is deep. IMHO, principle of least knowledge (A.K.A. [The Law of Demeter](https://en.wikipedia.org/wiki/Law_of_Demeter)) is good practice.